server/src/test/java/org/apache/kafka/server/ReconfigurableQuorumIntegrationTest.java

```diff
@@ -218,6 +218,82 @@ public void testControllersAutoJoinStandaloneVoter() throws Exception {
+                controllerNode.id() == TestKitDefaults.CONTROLLER_ID_OFFSET ?
+                        oldDirectoryId : controllerNode.metadataDirectoryId()
+            );
+        }
+
+        try (KafkaClusterTestKit cluster = new KafkaClusterTestKit.Builder(nodes).
+                setConfigProp(QuorumConfig.QUORUM_AUTO_JOIN_ENABLE_CONFIG, true).
+                setInitialVoterSet(initialVoters).
+                build()
+        ) {
+            cluster.format();
+            cluster.startup();
+            try (var admin = Admin.create(cluster.clientProperties())) {
+                // Static voter set is initialized
+                TestUtils.retryOnExceptionWithTimeout(30_000, 100, () -> {
+                    Map<Integer, Uuid> voters = findVoterDirs(admin);
+                    assertEquals(Set.of(3000, 3001, 3002), voters.keySet());
+                    for (int replicaId : new int[] {3000, 3001, 3002}) {
+                        assertEquals(nodes.controllerNodes().get(replicaId).metadataDirectoryId(), voters.get(replicaId));
+                    }
+                });
+
+                // Remove 3002 from voter set
+                TestUtils.retryOnExceptionWithTimeout(30_000, 100, () -> {
+                    Map<Integer, Uuid> voters = findVoterDirs(admin);
+                    if (!voters.containsKey(3002)) {
+                        // if there are no node 3002, it should be removed
+                        return;
+                    }
+
+                    admin.removeRaftVoter(3002, voters.get(3002)).all().get();
+                    assertEquals(Set.of(3000, 3001), voters.keySet());
+                    for (int replicaId : new int[] {3000, 3001}) {
+                        assertEquals(nodes.controllerNodes().get(replicaId).metadataDirectoryId(), voters.get(replicaId));
+                    }
+                });
+
+
+                // do not join the voter set in the next twenty seconds
+                for (int i = 0; i < 20; ++i) {
+                    TestUtils.retryOnExceptionWithTimeout(30_000, 100, () -> {
+                        TestUtils.retryOnExceptionWithTimeout(30_000, 100, () -> {
+                            Map<Integer, Uuid> voters = findVoterDirs(admin);
+                            if (!voters.containsKey(3002)) {
+                                // if there are no node 3002, it should be removed
+                                return;
+                            }
+
+                            admin.removeRaftVoter(3002, voters.get(3002)).all().get();
+                            assertEquals(Set.of(3000, 3001), voters.keySet());
+                            for (int replicaId : new int[] {3000, 3001}) {
+                                assertEquals(nodes.controllerNodes().get(replicaId).metadataDirectoryId(), voters.get(replicaId));
+                            }
+                        });
+                    });
+                    Thread.sleep(1000);
+                }
```

Comment:
This test doesn't make sense to me.
1. `do not join the voter set in the next twenty seconds`: The 20 seconds is not true because the `retryOnExceptionWithTimeout` could take up to 30 seconds one time.
2. Why should we use 2 `retryOnExceptionWithTimeout` here? 
3. The test is not testing what we expected. 
  3.1. What are you asserting here? I can only see you assert the node 3000, 3001 are in the voters, but we want to test 3002, right? If 3002 is in the voters, the test can still be passed!?
  3.2. If you want to test remove voters 20 times, then why do we need L257~L269 above? It's only doing the 21th time.
  3.3. Even if the above comments are all addressed, the test is still unreliable. The test can still be passed as long as the auto-join is done after the assertion.

Reply:
We can test won't auto join the removed node by:

1. Remove a controller node, ex: 3002.
2. Verify the highwatermark in controller metadata log is advancing by:
`cluster.raftManagers().get(3002)..client().highWatermark()`
The high watermark increasing means the follower is fetching the data from the leader, and because we auto join the controller node before fetching from the leader, if the node can be auto joined, it'll be already added.
3. Verify the voters are still 3000, 3001, without 3002.

So, we don't need the mysterious 20 times loop to do the voter removal.

What do you think?